### DISCUSSION TITLE
다중분류모델

### DISCUSSION THREAD

**Discussion by KIBEOMP** [Pleasant-riot/Lung-Disease-Detection#25] [Show and tell]
Inception V3

Inception V3 딥러닝 모델의 Metric 평가에 대한 결과값은
정확도는 94.80%, 정밀도는 97.24%, 재현율은 94.00%, F1 스코어는 95.59의 결과값을 나타내었다.
최종 에포크의 정확도는 학습 모델링의 경우 94.91%, 검증 모델링은 89.68%의 정확도를 나타내었다.
손실함수 값의 평가는 학습 모델링은 1.127%, 검증 모델링은 4.603%의 손실함수 값을 나타내었다.
이러한 결과로 Inception V3 딥러닝 모델은 흉부영상 데이터의 특징 추출 및 분류에 있어 매우 우수한 딥러닝 모델이며 학습상태 또한 매우 우수하다고 평가
model- https://www.koreascience.or.kr/article/JAKO202035066033782.pa1ff8ge

CNN 

https://www.analyticsvidhya.com/blog/2020/09/pneumonia-detection-using-cnn-with-implementation-in-python/
https://dining-developer.tistory.com/26
https://www.kaggle.com/kmader/train-simple-xray-cnn
https://medium.com/analytics-vidhya/detecting-anomalies-in-x-ray-using-cnn-1e4c2e49f23a
	-https://github.com/snehabhatt/MachineLearningProjects/blob/master/XRAY-Submit/Working_With_Chest_XRay_Images-2.ipynb

MobileNet V2

https://www.mdpi.com/2076-3417/11/6/2751/htm

Saliency Map

https://junstar92.tistory.com/153

추후 업데이트 예정

**Comment by SHEYCHAE** [comment]
# Inception V3 
- Inception V3는 노드간의 연결을 줄이고 행렬 연산은 Dense 연산을 하는 구조
- 기존 Inception모델과의 차이점: stem layer의 7x7 컨볼루션 연산을 3x3 연산 세 번 하는 것으로 대체 / 최적화 알고리즘으로 RMSProp사용
( 이전 딥러닝 모델의 경우 layer 증가에 따라 연산량도 함께 많아져 학습 시간 길어졌지만, 얘는 연산 복잡도를 줄이고 파라미터 수 적게 사용해서 **학습속도 향상**됨)
- Average Pooling 전의 Inception module을 더 넓게 형성하여 같은 입력값으로도 다양한 연산을 가능하게함으로써 차원을 줄이는데 발생하는 **정보 손실을 방지**
- Label smoothing 으로 데이터 정규화 작업을 통해 **과적합 방지**

**Comment by SHEYCHAE** [comment]
# MEMO
- 딥러닝이미지 처리분야  in CNN :: AlexNet, VGG, ResNet, InceptionNet 등...

![image](https://user-images.githubusercontent.com/77530039/143486460-f6f889f2-c02a-42d1-bb08-0caf3fd57b4e.png)
classification/ detection/segmentation
- 분류 : 이미지와 라벨을 학습
- 탐지(detection): disease 있는 부위에 바운딩박스
- segmentation: 픽셀위치를 학습해서 픽셀단위로 질병 추론함

**Reply by SHEYCHAE** [reply]
EfficientNet B7

**Reply by SHEYCHAE** [reply]
https://www.youtube.com/watch?v=5sXQGH9I2VM
https://hoya012.github.io/blog/EfficientNet-review/

**Comment by SHEYCHAE** [comment]
# EfficientNet B7 모델의 Noisy student ImageNet Pretrained 가중치를 기반으로 Multi GAP 형식의 분류 모델 

- EfficientNet B7 : Compound Scailing의 최적값을 찾아 적용하는 모델
- FPN (Feature Pyramid Network)은  다중 스케일 특징을 추출하는 구조 - FPN을 모방한 Multi Gap(Global Average Pooling) 구조 사용

#### multigap 구조 
<img src="https://user-images.githubusercontent.com/77530039/143489928-701e1b24-f49e-4857-b50d-9a16dbfe54dd.png" width="500" height="300"/>

layer별 특징을 추출하여 GAP로 압축한 후 연결하는 다중 구조
FPN의 많은 기능을 사용하는 장점 , 각 가능을 FC 구조로 직접 변환하는 대신 각 기능을 GAP로 압축하여 과부하의 위험을 줄이고 연산량을 줄일 수 있음
<img src="https://user-images.githubusercontent.com/77530039/143489952-d2bb2de0-fd8c-4211-b5d4-5d185b57c6bd.png" width="500" height="300"/>

Multi GAP으로 획득하고 Concatenete Layer와 병합된 기능은 512FC(Fully Connected Layer)를 두 번 거친 후 Softmax기능을 사용해 다중 클래스로 분류됨 
각 클래스는 질병명으로 나뉨. 
Encoder + Decoder 결합한 전체 모델 구조 (아래)
<img src="https://user-images.githubusercontent.com/77530039/143490148-12cab0dd-5808-4429-923b-09784233316e.png" width="500" height="300"/>

EfficientNet B7은 크기가 커서 과적합 위험도 높음. 그래서 FC를 통과하는 벡터에 L2 regularization 해줘야댐

----------------------------
클래스별로 데이터가 불균형할 경우 특정 클래스에 지나치게 편향된 학습 가능성 높음 --> 언더샘플링/오버샘플링 ! 
또는 클래스별로 업데이트 가중치를 다르게 .... (데이터가 많은 클래스에 낮은 가중치 , 적은 클래스에 높은 가중치 )
데이터 증강: Rand Augment 적용: 최대 수 N=2, 최대강도 M=28


--------------------------------------------
<img src="https://user-images.githubusercontent.com/77530039/143490873-f4a3427e-4615-45d2-a4af-999deb3f32bb.png" width="500" height="300"/>
<img src="https://user-images.githubusercontent.com/77530039/143491053-9f520be7-ae35-4282-88ba-4023aad9fb15.png" width="500" height="300"/>
<img src="https://user-images.githubusercontent.com/77530039/143491079-7b13ab22-e61d-43f1-8765-5d3008f7639e.png" width="500" height="300"/>


대충 EfficientNet B7 가중치 + Preprocessing + MultiGap 사용하면 가장 정확도 높다는 뜻

**Reply by SHEYCHAE** [reply]
뭔지 궁금하져?

**Reply by SHEYCHAE** [reply]
[applsci-11-09289-v3.pdf](https://github.com/Pleasant-riot/Pneumonia-Detection/files/7604909/applsci-11-09289-v3.pdf)
바로바로 2021년 10월에 나온 따끈따끈 논문입니당.. 근데 영어로 되어있어서 대충 해석해 놓은게 본문내용이에요 (정말 대충 해석함)
muti-class classiifcation of lung diseases << 바로바로 우리가 필요한 것,,,

**Reply by SHEYCHAE** [reply]
이 연구에서는 Segmentation 안하고 바로 전처리+ 분류해버린거같은데 
세그맨테이션 + 전처리 + (여기서도 했던) 데이터 증강  + 화질개선(가능하면) +이 연구에서 사용한 다중분류기법 
하면 소아xray에서도 정확도 높일 수 있을까요?

그리고 만약 소아 xray가 우리가 본 것 처럼.. 전신이 나온 xray가 많다면 segmentation으로만으로도 그런 xray의 폐를 분리해낼 수 있는지 내일 학습해보기로한거에 test해보고 
만약 안된다면..  yolo같은걸로 object detection해서 폐부분 학습해서 바운딩으로 뽑아낸 다음에 거기서 segmentation을 해야할지 의논해봅시다~

**Reply by SHEYCHAE** [reply]
<img src="https://user-images.githubusercontent.com/77530039/143494225-d8e07e56-1ea7-420e-8e9b-989849a7e3ad.png" width="300" height="300"/>


논문에 EfficientNet이라는게 나오는데 처음 들어봐서 검색을 좀 해봤어요
https://www.youtube.com/watch?v=5sXQGH9I2VM
https://hoya012.github.io/blog/EfficientNet-review/
여기에 친절하게 한국말로 설명되어 있지만 , 더 간단하게 설명하자면 

model을 만들 때 모델의 성능을 끌어올리는 방법 즉, model scaling방법에는 
Width Scaliing (filter +) / Depth Scaling ( layer +) / Resolution Scaling (input 해상도+)  가 있음 
여기서 EfficientNet은 이 **세 가지 scaling 방식을 모두 동시에 고려하여 최적의 비율을 찾아주는** 모델이에요

------------------------------------

이 논문에서는 EfficientNet B7의 Noisy Student로 ImageNet에서 학습한 weight를 이용하여 fine tuning learning을 수행하고 , Multi GAP 구조를 이용하여 각 layer 의 특징을 최대한 활용하였다네요


* 참고 *
Noisy student (EfficientNet)  : https://yeomko.tistory.com/42  ( 완벽히 이해못해서 낼 다시 리뷰하겠슴)
모델사이즈 작아서 inference속도는 빠른데 정확도는 높으니까 아주 효율적이라서 이름이 EfficientNet이래요 
B0~ B7까지 숫자가 올라갈수록 모델의 크기가 커집니다

**Reply by KIBEOMP** [reply]
와우..

**Comment by SHEYCHAE** [comment]
# Global Average Pooling (GAP) 이란...

Covolution Neural Network를 만들 때, Convolution 레이어와 Pooling을 사용함 
Pooling 중의 Global Average Pooling에 대해 알아보도록 하자 


### Pooling의 역할 
CNN에는 많은 convolution layer를 쌓기 때문에 필터의 수가 많음.
필터 수가 많으면 그만큼 feature map이 쌓인다는 뜻
즉, CNN의 차원은 매우 크다 
높은 차원을 다루려면 그 차원을 다룰 수 있는 많은수의 파라미터들을 필요
but , 파라미터가 너무 많아지면 학습 시 over fitting이 발생할 수 있음
So... 필터에 사용된 파라미터 수를 줄여서 차원을 감소시킬 방법이 필요! 
이런 역할을 CNN에서 해 주는 레이어가 pooling layer 라고 하네요

![image](https://user-images.githubusercontent.com/77530039/143573576-5e1b46f8-632d-4b14-af5b-50275b669959.png)

1.  Pooling 중 가장 많이 사용되는 Max Pooling을 이용하여 Pooling 설명
위 그림의 왼쪽을 보면 Max Pooling의 경우 빨간색 박스 안에서 가장 큰 값이 선택 되어 새로운 feature를 형성해요
빨간색 박스가 Pooling을 하는 영역이 됩니다.
박스의 height, width 각각이 2이기 때문에 사이즈는 2이고 박스의 이동 거리가 2이기 때문에 stride 도 2라고 합니다. 
(빨간색 박스의 꼭지점이 이동된 거리가 2인것을 참조)

Max pooling 처럼 가장 큰 값을 선택하는 이유는 
① 가장 큰 feature 값이 계산 시 가장 큰 영향을 주기 때문에 출력 값에 영향이 가장 크고 
② feature 즉, 특징을 가장 잘 나타내기 떄문
위 그림 예시와 같이 (2, 2) 영역을 보고 stride = 2의 Max Pooling을 한다면 feature의 크기는 반으로 줄어들면서 각 박스 영역의 최댓값만 선택되게 돼요

2. 최댓값을 취하지 않고 평균 값을 취하는 방법 == Average Pooling 
(feature의 모든 값들을 이용하면서 (평균) 하면서 feature의 크기를 줄여 차원을 감소시킴)


## Global Average Pooling

GAP(global average pooling)은 앞에서 설명한 Max(Average) Pooling 보다 더 급격하게 feature의 수를 줄입니다.
하지만 GAP의 목적은 앞에서 사용한 Pooling과 조금 달라요.
 **GAP의 목적은 feature를 1차원 벡터로 만들기 위함**입니다.

![image](https://user-images.githubusercontent.com/77530039/143577047-525cb9d4-2447-46e1-9dcd-e0008ba02bbb.png)

- 위 그림을 보면 같은 채널 (같은 색)의 feature들을 모두 평균을 낸 다음에 채널의 갯수(색의 갯수) 만큼의 원소를 가지는 벡터로 만들어요  --> 이런 방식으로 GAP는 (height, width, channel) 형태의 feature를 (channel,) 형태로 간단하게 만듦

- WHY 이렇게 극단적인 벡터 형태인 GAP을 사용하느냐
 :: GAP는 CNN + FC(Fully Connected) Layer에서 classifier인 **FC Layer를 없애기 위한 방법으로 도입**

--------------------------------
- FC Layer는 마지막 feature와 matrix 곱을 하여 feature 전체를 연산의 대상으로 삼아서 결과를 출력 --> 즉, feature가 이미지 전체를 함축하고 있다고 가정하면 이미지 전체를 보고 출력을 만들어요 
- But, FC layer를 classifier로 사용하는 경우 파라미터의 수가 많이 증가 / feature 전체를 matrix 연산하기 때문에 위치에 대한 정보도 사라지게 되는 단점이 존재
  + 더구나 FC Layer 사용 시 반드시 지정해 주어야 하는 FC layer의 사이즈로 인해 입력 이미지 사이즈 또한 그에 맞춰서 고정되어야 하는 단점이  있슴

--------------------------------
- 반면 GAP은 어떤 크기의 feature 라도 같은 채널의 값들을 하나의 평균 값으로 대체하기 때문에 벡터가 됨 
- So, **어떤 사이즈의 입력이 들어와도 상관이 없음.** 또한 단순히 (H, W, C) → (1, 1, C) 크기로 줄어드는 연산이므로 파라미터가 추가되지 않으므로 학습 측면에서도 유리, and 파라미터의 갯수가 FC Layer 만큼 폭발적으로 증가하지 않아서 over fitting 측면에서도 유리
- 따라서 GAP 연산 결과 1차원 벡터가 되기 때문에 **최종 출력에 FC Layer 대신 사용할 수 있어용**
- 경우에 따라서 FC layer와 같이 사용 되기도 합니다. FC layer에 전달하기 전에 GAP를 이용하여 차원을 줄여서 벡터로 만든 다음에 FC layer로 전달 하면 FC Layer에서 쉽게 사이즈를 맞출 수 있기 때문
